### PR TITLE
changed default image pull policy to if-not-present

### PR DESCRIPTION
**Pull request by droot**
`kpt fn` commands support command flag called `image-pull-policy` that determines image fetch behavior for KRM functions. Current default value is `AlwaysPull` which means `kpt` will always try to pull the image using `docker pull` even if the function image is present locally. In addition, current code also calls `docker inspect` to check if the image is present locally for `ifnotpresent`.

This PR does the following:
 - Changes the default value for `image-pull-policy` to `if-not-present` i.e. pull image only if it doesn't exists locally.
 - Removes the `docker pull` and `docker image inspect` calls instead relays the `image-pull-policy` flag to `docker run` command as is. This will eliminate two docker calls and thus, make the `fn render` and `fn eval` faster.
 - To support the above implementation, we had to filter out docker CLI command  (especially progress messages). This was needed because output from executing KRM function and `docker run` will be emitted on same `stderr` output stream.

### CODE REVIEW

File: e2e/testdata/fn-eval/missing-fn-image/.expected/config.yaml [hunk 1]
```diff
@@ -17,4 +17,4 @@ exitCode: 1
 image: gcr.io/kpt-fn/dne # non-existing image
 args:
   namespace: staging
-stdErr: 'Function image "gcr.io/kpt-fn/dne" doesn''t exist'
+stdErr: 'gcr.io/kpt-fn/dne:latest not found'
```
Comment (mengqiy) [inline e2e/testdata/fn-eval/missing-fn-image/.expected/config.yaml:20]: Why we have this change? I didn't find any changes related to this.

File: e2e/testdata/fn-render/fn-success-with-stderr/.expected/results.yaml [hunk 2]
```diff
@@ -5,6 +5,5 @@ metadata:
 exitCode: 0
 items:
   - image: gcr.io/kpt-fn/starlark:v0.2
-    stderr: |
-      function succeeded, reporting it on stderr
+    stderr: function succeeded, reporting it on stderr
```
Comment (mengqiy) [inline e2e/testdata/fn-render/fn-success-with-stderr/.expected/results.yaml:8]: Is this change necessary?

**Review by mengqiy — COMMENTED** [review]
Code changes LGTM

File: e2e/testdata/fn-eval/missing-fn-image/.expected/config.yaml (hunk 1 above)
Comment (droot) [reply e2e/testdata/fn-eval/missing-fn-image/.expected/config.yaml:20]: Earlier we were calling `docker pull` explicitly and capturing the not found error (look at the deleted code), so now that's gone, we relay the message from `docker run` as is. That error is little verbose because it's coming from docker, but I didn't find it much value in wrapping it up because the verbose message from docker is pretty clear.

File: e2e/testdata/fn-render/fn-success-with-stderr/.expected/results.yaml (hunk 2 above)
Comment (droot) [reply e2e/testdata/fn-render/fn-success-with-stderr/.expected/results.yaml:8]: This is needed because we are stripping away `\n` from the last line returned by the function when sanitizing the output from the function. It's because we are using `bufio.Scanner` to process the output. I might be able to work around it, but didn't think it matters. Do you think of it being a problem ?

Comment (mengqiy) [reply e2e/testdata/fn-render/fn-success-with-stderr/.expected/results.yaml:8]: I don't think that's a problem.

File: internal/fnruntime/container.go [hunk 3]
```diff
@@ -250,3 +194,48 @@ func (e *ContainerImageError) Error() string {
 		"Error: Function image %q doesn't exist remotely. If you are developing new functions locally, you can choose to set the image pull policy to ifNotPresent or never.\n%v",
 		e.Image, e.Output)
 }
+
+// filterDockerCLIOutput filters out docker CLI messages
+// from the given buffer.
+func filterDockerCLIOutput(in io.Reader) string {
```
Comment (sdowell) [inline internal/fnruntime/container.go:200]: What is the motivation for filtering the docker output from the buffer?

I haven't encountered this use case before, but I expected there would be a more elegant way to do this (e.g docker cli flag, I/O redirection, etc)

File: internal/fnruntime/container.go [hunk 4]
```diff
@@ -250,3 +194,48 @@ func (e *ContainerImageError) Error() string {
...
+//  "aafbf7df3ddf: Pulling fs layer"
+//  "aafbf7df3ddf: Verifying Checksum"
+//  "aafbf7df3ddf: Download complete"
+//  "6b759ab96cb2: Waiting"
+//  "aafbf7df3ddf: Pull complete"
+//  "Digest: sha256:c347e28606fa1a608e8e02e03541a5a46e4a0152005df4a11e44f6c4ab1edd9a"
+//  "Status: Downloaded newer image for gcr.io/kpt-fn/starlark:v0.3"
+//
+func isdockerCLIoutput(s string) bool {
+	if strings.Contains(s, "Already exists") ||
```
Comment (sdowell) [inline internal/fnruntime/container.go:228]: Minor nit: these could be made more strict by including the preceding colon
e.g. ": Already exists", ": Pulling fs layer", etc

**Comment by sdowell** [comment]
Code changes LGTM

File: internal/fnruntime/container.go (hunk 3 above)
Comment (droot) [reply internal/fnruntime/container.go:200]: so kpt invokes the functions using cmd.Exec() the tricky thing is the output stream (stdout/stderr) is shared between the docker command as well as the function itself. And at the kpt level, we want to process these outputs differently. I did investigate if docker provides a simple mechanism to control its progress output, but unfortunately couldn't find one, so had to settle for this hack.

I thought a lot about the maintenance aspect of this hack. docker is a very stable CLI, so I don't expect a huge churn in this code and also, this filtering is only relevant in case of error so even if the filtering is not perfect, it isn't going to effect the user experience much. And the upside of removing the two commands (`docker pull` and `docker image inspect`)  outweighs the downside of this hack, so went with this.

File: internal/fnruntime/container.go (hunk 4 above)
Comment (droot) [reply internal/fnruntime/container.go]: Good point. Done.

**Review by mengqiy — COMMENTED** [review]
LGTM
We can merge it after the fn catalog site part is done

File: internal/fnruntime/container.go (hunk 3 above)
Comment (sdowell) [reply internal/fnruntime/container.go:200]: Seems reasonable to me 🙂

**Review by sdowell — APPROVED** [review]
LGTM
